### ISSUE TITLE
Deep copy for directories

### ISSUE THREAD

**Issue by eichenberger** [goreleaser/nfpm#285]
Would it be possible to add something like directory as content type:
```
- src: install/lib/
  dst: /lib
  type: directory
```
I would like to package kernel modules which have normally some empty directories:
e.g:
```
install/lib/modules/5.4.79/file
```
The folders lib and modules are empty and the output with the configuration above is therefore:
```
/lib/file
```
However, I just would like to have a deep copy of the directory so that the output would be:
```
/lib/modules/5.4.79/file
```
This also happens when I disable globing. Therefore, I was thinking if we could add an additional type which forces this deep copy behaviour.

**Comment by caarlos0** [comment]
have you tried `src: install/lib/**/*`?

**Comment by eichenberger** [comment]
Hi @caarlos0 
Yes it's exactly the same when I do that. I think the problem is that the globing function is used to find the files even when globing is disabled (see https://github.com/goreleaser/nfpm/blob/master/files/files.go). For `src: install/lib/**/*` I somehow expect this behaviour because we want to have the shortest path possible. But without globing the whole path should be kept also for the destination (at least I would expect that). But maybe it's also wanted then the `type: directory` approach would also be okay.

**Comment by caarlos0** [comment]
just noticed the `type: directory`... it doesn't exist... 

you should probably leave it empty (and I think its probably warning about that in the logs)

**Comment by eichenberger** [comment]
Hi @caarlos0 
I didn't actually put `type: directory` into my config file. That's just the proposal to make a deep copy. If disabling globing would have the same effect that's also fine. 

Just as an example you can do the following:
```
mkdir -p a/b/c/d
echo hallo1 > a/b/c/d/test1
echo hallo2 > a/b/c/d/test2
```
Then use the following nfpm.yml file to create a deb package:
```
name: "foo"
arch: "amd64"
platform: "linux"
version: "v1.0.0"
section: "default"
priority: "extra"
replaces:
- foobar
provides:
- bar
depends:
- foo
- bar
recommends:
- whatever
suggests:
- something-else
conflicts:
- not-foo
- not-bar
maintainer: "John Doe <john@example.com>"
description: |
  FooBar is the great foo and bar software.
    And this can be in multiple lines!
vendor: "FooBarCorp"
homepage: "http://example.com"
license: "MIT"
disable_globbing: true
contents:
- src: a
  dst: /a
```
Run:
```
nfpm pkg -f nfpm.yaml -p deb
```
 The deb file will only include a file a/test1 and a file a/test2 which is in my opinion wrong (I disabled globbing). It should contain a/b/c/d/test1 and a/b/c/d/test2. However, if that behaviour is wanted then my proposal was to add a `type: directory` so that this use case would still work.

**Comment by djgilcrease** [comment]
@eichenberger 
> The deb file will only include a file a/test1 and a file a/test2 which is in my opinion wrong (I disabled globbing). It should contain a/b/c/d/test1 and a/b/c/d/test2. However, if that behavior is wanted then my proposal was to add a `type: directory` so that this use case would still work.

In my testing with your above example, it works as you would expect with no modifications to code. I am in the process of cleaning up some of the testing so will add an explicit example to the tests to cover this expectation. I will keep this open until I get the tests merged.

**Comment by eichenberger** [comment]
@djgilcrease  
If it's intended to work like this would it then be possible to add a new feature? Because, I just would like to have a 1:1 copy of the directory structure inside the package. I can understand that this is maybe not the normal use-case but it would be super helpful to just have the pendant to:
```
tar -czf ../data.tar.gz -C a .
```
It would also be okay if that is an additional type (as proposed e.g. directory).

**Comment by djgilcrease** [comment]
You can see with the test I added that is is doing a full copy of the directory https://github.com/goreleaser/nfpm/pull/292/files?file-filters%5B%5D=.dockerfile&file-filters%5B%5D=.go&file-filters%5B%5D=.pol&file-filters%5B%5D=.txt&file-filters%5B%5D=.yaml#diff-cd252b527f9f0199d5b4f6a1fecdd0f2f4037e8a0325fb46bc52e99208fa6594R69

So with globbing off
```yaml
contents:
- src: testdata/globtest
  dst: /bla
```
turns into
```yaml
contents:
- src: testdata/globtest/a.txt
  dst: /bla/a.txt
- src: testdata/globtest/nested/b.txt
  dst: /bla/nested/b.txt
- src: testdata/globtest/multi-nested/subdir/c.txt
  dst: /bla/multi-nested/subdir/c.txt
```

Which is what I would expect.

**Comment by eichenberger** [comment]
@djgilcrease 
I agree your test really works and does what I also expect. However, I found out how I can provoke the issue. If you delete files/testdata/globtest/a.txt and files/testdata/globtest/nested then the test will fail because c.txt will be in the wrong directory. I did:
```
rm files/testdata/globtest/a.txt
rm -rf files/testdata/globtest/nested
```
Then if you run the test and it will fail with the following error:
```
=== RUN   TestDeepPathsWithGlob
    files_test.go:64:
                Error Trace:    files_test.go:64
                Error:          Not equal:
                                expected: "/bla/multi-nested/subdir/c.txt"
                                actual  : "/bla/c.txt"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -/bla/multi-nested/subdir/c.txt
                                +/bla/c.txt
                Test:           TestDeepPathsWithGlob
--- FAIL: TestDeepPathsWithGlob (0.00s)
FAIL
```
I hope this helps

**Comment by djgilcrease** [comment]
yes, that helps, I will dig into why.

**Comment by daledude** [comment]
I've run into this as well.
```
My local dir tree:
drwxr-xr-x   etc/
drwxr-xr-x   etc/cni
drwxr-xr-x   etc/cni/net.d
-rw-r--r--   etc/cni/net.d/87-podman-bridge.conflist

nfpm.yaml:
src: ./etc/cni
OR
src: ./etc/cni/*
OR
src: ./etc/cni/**/*
```
All cause the below to be created which is ignoring the net.d subdir.
```
-rw-r--r--   /etc/cni/87-podman-bridge.conflist
```

**Comment by djgilcrease** [comment]
Added a fix and a test to verify in #292 

https://github.com/goreleaser/nfpm/blob/issue-285/files/files_test.go#L71
